### ISSUE TITLE
A question about the task-8

### ISSUE THREAD

**Issue by 14H034160212** [nuric/deeplogic#8]
Hi,

I am trying to solve the task 8 with several methods. But from the current result which did not show an evident improvement.

Firstly I only use train_task8 and valid_task8 as training and validation sets to train the model, but from the current train and valid accuracy which is around 59% looks like similar like the multi-task training result. 

The method I am trying to improve is I have created a task-13 which is a three-hop transitivity dataset, I put this dataset in the multi-task learning, but the valid accuracy of task-8 is about 60% which is not evident better but one strange thing is the valid accuracy of the three-hop transitivity is higher than 90%.

Here is an example of the three-hop transitivity:
`
xwb(Y,T):-mha(Y,O);w(O,L);g(L,T).
mha(xmca,l).
w(l,u).
w(cs,u).
zm(l,u).
g(u,g).
gloe(u,g).
mha(nwcj,cpm).
tlav(bs,cpm).
r(cpm,hzdn).
jz(hzdn,phip,k).
j(O):--toy(O).
f(B,D):-rm(D,B).
? xwb(xmca,g). 1
? xwb(bs,phip). 0
`

I also generate the hard version:
`
h(Y,W):-knyggekpyul(Y,V);xnocvwutbd(V,I);nrlnqlxe(I,W).
knyggekpyul(lcskej,paqehvoufaf).
nwdvl(lcskej,paqehvoufaf).
xnocvwutbd(paqehvoufaf,vattisafpxdq).
nrlnqlxe(vattisafpxdq,jwxn).
lbugeoyy(vattisafpxdq,jwxn).
knyggekpyul(jtzawudjzlzw,mtsvkcmgvr).
vlkedj(jtzawudjzlzw,mtsvkcmgvr).
xnocvwutbd(quddtixtset,ns).
nex(T,T).
ltqnkbkbbo(H,F):-xkpbz(F,H).
n(S):--y(S).
xbwh(P,O):-hyiqx(P,O).
jes(I,B):-stoixuu(B,I).
foh(O,Q):-amfrhpxio(O,Q).
lnvlmlpqcq(S,V):-xenfid(V,S).
jesgporb(B,M):-cn(M,B).
nr(Z,T):-qmxxuryrkly(Z,T).
nh(C,D):-tsbsyl(D,C).
rqacnbdkxzg(sorkc).
ph(N):--ntlfugdc(N).
? h(lcskej,jwxn). 1
? h(jtzawudjzlzw,zjsiijewzmt). 0
`

Besides, I find a small issue that there is semicolon to connect the predicate in the rule rather than comma which actually existing in the other synthetic datasets.

Here is the result:
<img width="885" alt="immediate experiment result" src="https://user-images.githubusercontent.com/23516191/90473973-595fd800-e178-11ea-999d-c5ad78918324.png">

One question is you see the difference between the hard version dataset and the easy version dataset is the hard version has more noise dataset and longer predicate, but actually the length should not be a important factor that it is just a symbol. So, I think word-level embedding might be a better way to use character-level embedding in this case. 

Therefore, have you seen any tree-based structure used in the symbolic task? I guess that might be a better way for the symbolic reasoning. I saw one classical paper which talks about the tree-LSTM (https://arxiv.org/pdf/1503.00075.pdf) and attentive recursive trees (https://arxiv.org/pdf/1811.02338.pdf).

**Comment by nuric** [comment]
Hello, it's great to see that you are exploring future work. Unfortunately, it would be difficult to comment on this issue from my perspective as it is beyond the scope of this repository and novel work.

The semi-colon was used to make easier to parse the logic rules afterwards. You can change it to any other character you like. By default, it is true that ',' comma is used in logic programs to specify a conjuction in the rule body.

Finally, there are many different ways you can compose a neural network architecture and again it depends on the new tasks you are trying to solve. I would encourage to research more along the lines of tree-LSTM and graph neural networks, a really nice paper explaining them is titled [Relational inductive biases, deep learning, and graph networks](https://arxiv.org/abs/1806.01261).
